src/main/java/frc/robot/SwerveModule.java

```diff
@@ -0,0 +1,108 @@
+package frc.robot;
+
+import com.ctre.phoenix.sensors.CANCoder;
+import com.ctre.phoenix.sensors.SensorTimeBase;
+import com.revrobotics.RelativeEncoder;
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import frc.robot.Constants.ModuleConstants;
+import frc.robot.Constants.DriveConstants;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.RobotController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class SwerveModule {
+    private final CANSparkMax driveMotor;
+    private final CANSparkMax steerMotor;
+    
+    private final RelativeEncoder driveEncoder;
+    private final RelativeEncoder steerEncoder;
+
+    private final PIDController steerPidController;
+
+    private final CANCoder absoluteEncoder;
+    private final boolean absoluteEncoderReversed;
+    private final double absoluteEncoderOffsetRad;
+
+    public SwerveModule(int driveMotorId, int steerMotorId, boolean driveMotorReversed, boolean steerMotorReversed,
+    int absoluteEncoderId, double absoluteEncoderOffset, boolean absoluteEncoderReversed) {
+    //absolute encoders
+    this.absoluteEncoderOffsetRad = absoluteEncoderOffset;
+    this.absoluteEncoderReversed = absoluteEncoderReversed;
+    absoluteEncoder = new CANCoder(absoluteEncoderId); //default is degrees per second
+    absoluteEncoder.configFeedbackCoefficient(ModuleConstants.ABSOLUTE_ENCODER_CPM_TO_RAD, "rad", SensorTimeBase.PerSecond); //convert to radians per second
+    //motors
+    driveMotor = new CANSparkMax(driveMotorId, MotorType.kBrushless);
+    steerMotor = new CANSparkMax(steerMotorId, MotorType.kBrushless);
+    //inverting motors
+    driveMotor.setInverted(driveMotorReversed);
+    steerMotor.setInverted(steerMotorReversed);
+    //motor encoders
+    driveEncoder = driveMotor.getEncoder();
+    steerEncoder = steerMotor.getEncoder();
+
+    //so that we can work with meters and radians instead of rotations
+    driveEncoder.getPosition();
+    driveEncoder.setPositionConversionFactor(ModuleConstants.DRIVE_ENCODER_ROT_TO_METERS);
+    driveEncoder.setVelocityConversionFactor(ModuleConstants.DRIVE_ENCODER_RPM_TO_METERS_PER_SEC);
+    steerEncoder.setVelocityConversionFactor(ModuleConstants.STEER_ENCODER_ROT_TO_RAD);
+    steerEncoder.setVelocityConversionFactor(ModuleConstants.STEER_ENCODER_RPM_TO_RAD_PER_SEC);
+    
+    steerPidController = new PIDController(ModuleConstants.P_STEER, 0, 0);
```

Comment:
Oh cool! Didn't know that. How do you implement a hardware PID?

Reply:
[Here's their documentation entry on it](https://docs.revrobotics.com/sparkmax/operating-modes/closed-loop-control)
[Here's their example of position closed loop control](https://github.com/REVrobotics/SPARK-MAX-Examples/tree/master/Java/Position%20Closed%20Loop%20Control)

The problem though is that you can't do hardware PID with the CANcoder, you would have to do it with the internal encoder, which has low resolution (I did see a team on the forums that claims they do this), or you could've done it with a mag encoder since it would plug straight into the controller.